src/integration/test-helpers.ts

```diff
@@ -123,3 +123,29 @@ export const waitUntilRecordsReady = async (
 
   return indexStats;
 };
+
+type Assertion = (result: any) => void;
+
+export const assertWithRetries = async (
+  asyncFn: () => Promise<any>,
+  assertions: Assertion[],
+  maxRetries: number = 5,
+  delay: number = 3000
+) => {
+  let attempts = 0;
+
+  while (attempts < maxRetries) {
+    try {
+      const result = await asyncFn();
+      assertions.forEach((assertion) => assertion(result));
+      return;
+    } catch (error) {
+      attempts++;
+      if (attempts < maxRetries) {
+        await sleep(delay);
+      } else {
+        throw error;
+      }
+    }
+  }
+};
```

Comment:
It might be nice to adjust the delay exponential backoff style. Five doublings would turn a 3 second delay into a `3*32 = 96` second delay. Total waiting possible would be `(3*(1 + 2 + 4 + 8 + 16 + 32)) = 189 seconds`. Seems reasonable if that's enough to make the test pass consistently.
 
Also, is there an overall jest timeout you need to worry about running into?

```
export const assertWithRetries = async (
  asyncFn: () => Promise<any>,
  assertions: Assertion[],
  maxRetries: number = 5,
  initialDelay: number = 3000
) => {
  let attempts = 0;
  let delay = initialDelay;

  while (attempts < maxRetries) {
    try {
      const result = await asyncFn();
      assertions.forEach((assertion) => assertion(result));
      return;
    } catch (error) {
      attempts++;
      if (attempts < maxRetries) {
        await sleep(delay);
        delay *= 2; // Double the delay for exponential backoff
      } else {
        throw error;
      }
    }
  }
};
```

Reply:
Good call, agreed with increasing the delay each run, I think that can only help us.

As for the jest timeout, I believe we're currently configured at `100000` ms globally: https://github.com/pinecone-io/pinecone-ts-client/blob/757185042c5c4703b5a4736f4e9836fcf1d65c6f/jest.config.js#L15C19-L15C19

I think this should be more than enough. Given I'm checking `attempts < maxRetries` I think there's only 4 doublings. I think bumping up to `attempts <= maxRetries` gives us one more run, and we should still be under the global timeout:

3 seconds
6 seconds
12 seconds
24 seconds
48 seconds
= 93 seconds

100,000 ms = 100 seconds